File: cpp_10-99/71.cc
// C++
// leetcode 71
// https://leetcode-cn.com/problems/simplify-path/

class Solution {
private:
    
public:
    string simplifyPath(string path) {
        vector<string> stk;

        int start = 0, idx = start;
        while(idx < path.size()) {
            // 寻找使得path[start, idx)为下一个目录
            start = idx;
            while(start < path.size() && path[start] == '/')
                start++;
            idx = start;
            while(idx < path.size() && path[idx] != '/')
                idx++;
            if(start == path.size())
                break;
            // 至此，path[start, idx)为下一个符合条件的目录

            // 判断这一次的目录，并作出相应操作
            if(idx == start + 1 && path[start] == '.') {
                // "."  当前目录，啥也不做
            } else if(idx == start + 2 && path[start] == '.' && path[start + 1] == '.') {
                // ".." 返回上一级目录，如果此时非根目录，则stk出栈一次
                if(!stk.empty())
                    stk.pop_back();
            } else {
                // 是一个合格的目录，可以录入stk
                stk.push_back(path.substr(start, idx - start));
            }
        }
        // 构造结果的字符串
        string res = "";
        if(stk.size() == 0) {
            res = "/";
        } else {
            for(auto s : stk)
                res += "/" + s;
        }
        return  res;
    }
};

File: Go/Go_10-99/71.go
// Go
// leetcode 71
// https://leetcode-cn.com/problems/simplify-path/

func simplifyPath(path string) string {
	n := len(path)
	stk := []string{}
	for i := 0; i < n;  {
		for i < n && path[i] == '/' {
			i++
		}
		j := i
		for j < n && path[j] != '/' {
			j++
		}
        if i == j {
            break
        }
		cur := path[i : j]
        // fmt.Println(cur)
		switch cur {
		case ".":
		case "..":
			if len(stk) != 0 {
				stk = stk[0: len(stk) - 1]
			}
		default:
			stk = append(stk, cur)
		}
        i = j
	}

	if len(stk) == 0 {
		return "/"
	}

	res := ""
	for _, s := range stk {
		res += "/" + s
	}
	return res
}